### PR TITLE
Onion tool and sphinx minor cleanups

### PR DESCRIPTION
**Pull request by rustyrussell**
Prep work for messages.

Changelog-None

### CODE REVIEW

File: devtools/onion.c [hunk 1]
```diff
@@ -174,16 +169,17 @@ static void do_decode(int argc, char **argv, const u8 assocdata[ASSOC_DATA_SIZE]
...
 {
-	if (!hex_decode(arg, strlen(arg), assocdata, ASSOC_DATA_SIZE))
+	*assocdata = tal_hexdata(NULL, arg, strlen(arg));
+	if (!*assocdata)
 		return "Bad hex string";
 	return NULL;
 }
 
-static void opt_show_ad(char buf[OPT_SHOW_LEN], const u8 *assocdata)
+static void opt_show_ad(char buf[OPT_SHOW_LEN], u8 *const *assocdata)
```
Comment (niftynei) [inline devtools/onion.c]: ooh this is fancy. what's a `*const`?

File: common/onion.c [hunk 2]
```diff
@@ -176,10 +177,15 @@ static bool pull_payload_length(const u8 **cursor,
 	 *   case the `hop_payload_length` is equal to the numeric value of
 	 *   `length`.
 	 */
-	if (*len > 1) {
+	if (!has_realm || *len > 1) {
```
Comment (niftynei) [inline common/onion.c]: what if `*len == 1`?
```suggestion
	if (!has_realm || *len > 0) {
```

Comment (niftynei) [reply common/onion.c]: oh i just saw this comment ... ignore this.

```
 identified by any length over **`1`**
```

**Comment by niftynei** [comment]
ACK https://github.com/ElementsProject/lightning/pull/3587/commits/f192a23a429c1b3c2e0a4b5323f824a512a11557

figured out where the `1` case is :)

File: devtools/onion.c (hunk 1 above)
Comment (cdecker) [reply devtools/onion.c]: I think `const u8 *` is a constant pointer (so the address it points to can't be modified) while `u8 *const *` is a pointer to a const (so the pointee cannot be modified). I've never seen this before either, so maybe @rustyrussell can shed some light on this?

File: common/onion.h [hunk 3]
```diff
@@ -39,6 +39,7 @@ u8 *onion_final_hop(const tal_t *ctx,
  * onion_payload_length: measure payload length in decrypted onion.
  * @raw_payload: payload to look at.
  * @len: length of @raw_payload in bytes.
+ * @has_realm: used for HTLCs, where first byte 0 is magical.
```
Comment (cdecker) [inline common/onion.h]: `0` and `1` are both special

**Pull request merged**